File: Factory Method/Wave.java
import java.util.ArrayList;

public class Wave {

    private ArrayList<Shape> shapes;

    public Wave() {
        shapes = new ArrayList<>();
    }

    public void add(Shape shape)
    {
        shapes.add(shape);
    }

    public final void print() {
        for (Shape ss : shapes)
            ss.print();
    }
};

File: Prototype/Wave.cpp
#include "Wave.h"
#include "Shape.h"

void Wave::add(Shape* shape)
{
    shapes.push_back(shape);
}

void Wave::print() const
{
    for (int i = 0; i < shapes.size(); ++i)
    {
        shapes.at(i)->print();
    }
}